--- iputils/rdisc.c.alias	2008-11-07 11:59:51.000000000 +0100
+++ iputils/rdisc.c	2008-11-07 12:27:02.000000000 +0100
@@ -1213,22 +1213,37 @@
 int
 join(int sock, struct sockaddr_in *sin)
 {
-	int i;
+	int i, j;
 	struct ip_mreqn mreq;
+	int joined[num_interfaces];
+	int skip;
 
 	if (isbroadcast(sin))
 		return (0);
 
+	for (i = 0; i < num_interfaces; i++) {
+		joined[i] = 0;
+	}
 	mreq.imr_multiaddr = sin->sin_addr;
 	for (i = 0; i < num_interfaces; i++) {
 		mreq.imr_ifindex = interfaces[i].ifindex;
 		mreq.imr_address.s_addr = 0;
-
+		skip = 0;
+		for (j = 0; j < i; j++) {
+			if (joined[j] == mreq.imr_ifindex)
+			{
+				skip = 1;
+				break;
+			}
+		}
+		if (skip)
+			continue;
 		if (setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP,
 			       (char *)&mreq, sizeof(mreq)) < 0) {
 			logperror("setsockopt (IP_ADD_MEMBERSHIP)");
 			return (-1);
 		}
+		joined[i] = mreq.imr_ifindex;
 	}
 	return (0);
 }
