diff -urp iputils-s20071127/ping.c iputils-new/ping.c
--- iputils-s20071127/ping.c	2014-11-18 14:46:35.733874105 +0100
+++ iputils-new/ping.c	2014-11-18 14:34:39.638178153 +0100
@@ -258,7 +258,7 @@ main(int argc, char **argv)
 		} else {
 			rc = idna_to_ascii_lz (target, &idn, 0);
 			if (rc == IDNA_SUCCESS)
-				hp = gethostbyname (idn);
+				hp = gethostbyname2 (idn, AF_INET);
 			else {
 				fprintf(stderr, "ping: IDN encoding of '%s' failed with error code %d\n", target, rc);
 				exit(2);
diff -urp iputils-s20071127/tracepath.c iputils-new/tracepath.c
--- iputils-s20071127/tracepath.c	2014-11-18 14:46:35.737874092 +0100
+++ iputils-new/tracepath.c	2014-11-18 14:45:32.110077934 +0100
@@ -319,9 +319,9 @@ main(int argc, char **argv)
 		base_port = atoi(p+1);
 	} else
 		base_port = 44444;
-	he = gethostbyname(argv[0]);
+	he = gethostbyname2(argv[0], AF_INET);
 	if (he == NULL) {
-		herror("gethostbyname");
+		herror("gethostbyname2");
 		exit(1);
 	}
 	memcpy(&target.sin_addr, he->h_addr, 4);
