diff -ur iputils-s20071127.orig/arping.c iputils-s20071127/arping.c
--- iputils-s20071127.orig/arping.c	2013-08-30 11:20:00.181750396 +1000
+++ iputils-s20071127/arping.c	2013-08-30 11:20:27.452749409 +1000
@@ -202,16 +202,22 @@
 	ts_o.tv_sec = timeout;
 	ts_o.tv_nsec = 500 * 1000000;
 
-	if (count-- == 0 || (timeout && timespec_later(&ts_s, &ts_o)))
+	if (timeout && timespec_later(&ts_s, &ts_o))
 		finish();
 
 	timespec_sub(&ts, &last, &ts_s);
 	ts_o.tv_sec = 0;
 
 	if (last.tv_sec==0 || timespec_later(&ts_s, &ts_o)) {
+		if (!timeout && (sent == count))
+			finish();
 		send_pack(s, src, dst,
 				  (struct sockaddr_ll *)&me, (struct sockaddr_ll *)&he);
-		if (count == 0 && unsolicited)
+		if ((sent == count) && unsolicited)
+			/* We usually wait for an extra iteration
+			 * after sending the last request to see if we
+			 * get a reply, but we don't need to in
+			 * unsolicited mode */
 			finish();
 	}
 	alarm(1);
@@ -321,6 +327,8 @@
 		fflush(stdout);
 	}
 	received++;
+	if (timeout && (received == count))
+		finish();
 	if (FROM->sll_pkttype != PACKET_HOST)
 		brd_recv++;
 	if (ah->ar_op == htons(ARPOP_REQUEST))
@@ -633,8 +641,6 @@
 		sigaddset(&sset, SIGINT);
 		sigprocmask(SIG_BLOCK, &sset, &osset);
 		recv_pack(packet, cc, &from[0]);
-		if(received == count)
-			exit(0);
 		sigprocmask(SIG_SETMASK, &osset, NULL);
 	}
 }
Only in iputils-s20071127: arping.c~
Only in iputils-s20071127: arping.c.exit-conditions
diff -ur iputils-s20071127.orig/doc/arping.sgml iputils-s20071127/doc/arping.sgml
--- iputils-s20071127.orig/doc/arping.sgml	2007-11-27 10:57:27.000000000 +1000
+++ iputils-s20071127/doc/arping.sgml	2013-08-30 11:20:16.363749810 +1000
@@ -58,8 +58,8 @@
 Stop after sending <replaceable/count/ ARP REQUEST
 packets. With 
 <link linkend="arping.deadline"><replaceable/deadline/</link>
-option, <command/arping/ waits for
-<replaceable/count/ ARP REPLY packets, until the timeout expires.
+option, instead wait for
+<replaceable/count/ ARP REPLY packets, or until the timeout expires.
   </para></listitem>
  </varlistentry>
 
Only in iputils-s20071127/doc: arping.sgml.exit-conditions
