diff -up iputils-s20071127/Makefile.ping_flags iputils-s20071127/Makefile
--- iputils-s20071127/Makefile.ping_flags	2010-09-06 08:13:35.521374681 +0200
+++ iputils-s20071127/Makefile	2010-09-06 08:14:19.914365458 +0200
@@ -29,10 +29,10 @@ all: $(TARGETS)
 tftpd: tftpd.o tftpsubs.o
 
 ping: ping.o ping_common.o
-	$(CC) $(CFLAGS) ping.o ping_common.o -lidn -o ping
+	$(CC) $(CFLAGS) $(LDFLAGS) ping.o ping_common.o -lidn -o ping
 
 ping6: ping6.o ping_common.o
-	$(CC) $(CFLAGS) ping6.o ping_common.o -o ping6
+	$(CC) $(CFLAGS) $(LDFLAGS) ping6.o ping_common.o -o ping6
 
 ping.o ping6.o ping_common.o: ping_common.h in6_flowlabel.h
 tftpd.o tftpsubs.o: tftp.h
