--- iputils-s20071127/arping.c	2013-05-22 08:43:39.294127882 +0200
+++ iputils-s20071127-new/arping.c	2013-05-22 08:47:42.578378153 +0200
@@ -565,13 +565,17 @@ main(int argc, char **argv)
 
                 if ((brdcast_val = read_sysfs_broadcast(brdcast_path)) == NULL) {
                         perror("sysfs read broadcast value");
-                        exit(2);
-                }
-                for (ch=0; ch<he[0].sll_halen; ch++) {
-                        he[0].sll_addr[ch] = strtol(brdcast_val + (ch*3), &next_ch, 16);
-                }
+                        fprintf(stderr, "falling back to default broadcast value\n");
+                        for (ch=0; ch<he[0].sll_halen; ch++) {
+                                he[0].sll_addr[ch] = 0xff;
+                        }
+                } else {
+                        for (ch=0; ch<he[0].sll_halen; ch++) {
+                                he[0].sll_addr[ch] = strtol(brdcast_val + (ch*3), &next_ch, 16);
+                        }
 
-                free(brdcast_val);
+                        free(brdcast_val);
+                }
 	}
 
 	if (!quiet) {
